src/core/lib/event_engine/cf_engine/cf_engine.cc

```diff
@@ -0,0 +1,139 @@
+    grpc_core::MutexLock lock(&mu_);
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
+      for (auto handle : known_handles_) {
+        gpr_log(GPR_ERROR,
+                "WindowsEventEngine:%p uncleared TaskHandle at shutdown:%s",
+                this, HandleToString(handle).c_str());
+      }
+    }
+    GPR_ASSERT(GPR_LIKELY(known_handles_.empty()));
+    timer_manager_.Shutdown();
+  }
+  executor_->Quiesce();
+}
+
+absl::StatusOr<std::unique_ptr<EventEngine::Listener>>
+CFEventEngine::CreateListener(
+    Listener::AcceptCallback on_accept,
+    absl::AnyInvocable<void(absl::Status)> on_shutdown,
+    const EndpointConfig& config,
+    std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+CFEventEngine::ConnectionHandle CFEventEngine::Connect(
+    OnConnectCallback on_connect, const ResolvedAddress& addr,
+    const EndpointConfig& args, MemoryAllocator memory_allocator,
+    Duration timeout) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+bool CFEventEngine::CancelConnect(ConnectionHandle handle) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+bool CFEventEngine::IsWorkerThread() { GPR_ASSERT(false && "unimplemented"); }
+
+std::unique_ptr<EventEngine::DNSResolver> CFEventEngine::GetDNSResolver(
+    const DNSResolver::ResolverOptions& options) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+void CFEventEngine::Run(EventEngine::Closure* closure) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+void CFEventEngine::Run(absl::AnyInvocable<void()> closure) {
+  GPR_ASSERT(false && "unimplemented");
+}
+
+EventEngine::TaskHandle CFEventEngine::RunAfter(Duration when,
+                                                EventEngine::Closure* closure) {
+  return RunAfterInternal(when, [closure]() { closure->Run(); });
+}
+
+EventEngine::TaskHandle CFEventEngine::RunAfter(
+    Duration when, absl::AnyInvocable<void()> closure) {
+  return RunAfterInternal(when, std::move(closure));
+}
+
+bool CFEventEngine::Cancel(TaskHandle handle) {
+  grpc_core::MutexLock lock(&mu_);
+  if (!known_handles_.contains(handle)) return false;
+  auto* cd = reinterpret_cast<Closure*>(handle.keys[0]);
+  bool r = timer_manager_.TimerCancel(&cd->timer);
+  known_handles_.erase(handle);
+  if (r) delete cd;
+  return r;
+}
+
+EventEngine::TaskHandle CFEventEngine::RunAfterInternal(
+    Duration when, absl::AnyInvocable<void()> cb) {
+  auto when_ts = ToTimestamp(timer_manager_.Now(), when);
+  auto* cd = new Closure;
+  cd->cb = std::move(cb);
+  cd->engine = this;
+  EventEngine::TaskHandle handle{reinterpret_cast<intptr_t>(cd),
+                                 aba_token_.fetch_add(1)};
+  grpc_core::MutexLock lock(&mu_);
+  known_handles_.insert(handle);
+  cd->handle = handle;
+  GRPC_EVENT_ENGINE_TRACE("CFEventEngine:%p scheduling callback:%s", this,
+                          HandleToString(handle).c_str());
+  timer_manager_.TimerInit(&cd->timer, when_ts, cd);
+  return handle;
+}
```

Comment:
just curious -- probably not worth it, but since a decent amount of these code are pretty much the same as in the posix engine, is there an opportunity for code reuse?

Reply:
Maybe

Reply:
Ah ok. Could you maybe evaluate and offer your thoughts here?

And if it's not worth to do it right now in this PR, we could submit as is. But keep it in mind as we add further code, and adopt code reuse if the amount of code warrants the additional complexity.

Thanks!

Reply:
Discussed offline. Timer is the only place where there's potential to reuse, but not the other components. So it's probably not worth the additional complexity.